docs(blog): add kubeflow setup guide (#1371)
glasskube/glasskube: website/blog/2024-10-29-kubeflow-setup/index.mdx

---
slug: kubeflow-setup-guide
title: Getting Started with Cloud-native AI using Kubeflow
description: |
  In this article, I would like to introduce to you Kubeflow: A complete and cloud-native platform that simplifies AI operations.
  Join me in setting up Kubeflow on GKE for your organization and get started with cloud-native AI today.
authors:
  - kosmoz
tags: [guide, ai, mlops, kubernetes, eks, gcp, azure]
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

In this article, I would like to introduce to you Kubeflow: A complete and cloud-native platform that simplifies AI operations.
Join me in setting up Kubeflow on GKE for your organization and get started with cloud-native AI **today**.

{/* truncate */}

## Prerequisites

> - Up-to-date version of `git`, `kubectl` and `kustomize` installed

Reviewer: Might be interesting to include a backlink somewhere to link to the Kustomize guide: https://glasskube.dev/blog/patching-with-kustomize/

Author: That's a great idea. I think I'll add one closer to the end where the article suggests that people should take this as a basis to build their own kubeflow from

- Up-to-date version of `gcloud` and the `gke-gcloud-auth-plugin` component installed and initialized
- A Google Cloud project with sufficient funds to host the GKE cluster
- A domain name and the capabilities to create DNS records for that domain

## Background

### AI on Kubernetes

Over the course of the last decade, Kubernetes has established itself as one of the foundational pillars of modern cloud computing.

> Countless enterprises and organizations have adopted it as a platform to serve their internal and external IT infrastructure, benefiting from it's declarative management, self-healing and scaling capabilities.

Reviewer 2: ```suggestion
Countless enterprises and organizations have adopted it as a platform to serve their internal and external IT infrastructure, benefiting from its declarative management, self-healing and scaling capabilities.
```

But the advance of Artificial Intelligence (AI) in recent years has presented the community with several challenges.

While containerization, workload scheduling, horizontal scaling and the general abstraction for compute resources are all relevant to AI engineers,
Kubernetes was designed first and foremost with web applications in mind, and not machine learning tasks and inference serving.
For this reason, the primitives Kubernetes provides for users to manage workloads have turned out to be unintuitive for individuals tasked with managing AI workloads if they don't already have a deep understanding of the system.

> Furthermore, AI workloads have very different performance characteristics compared to the web servers and database servers we are to.

Reviewer 2: ```suggestion
Furthermore, AI workloads have very different performance characteristics compared to the web servers and database servers we are used to.
```

> They are, generally, very resource demanding, strongly coupled and often short lived.

Reviewer 2: ```suggestion
They are, generally, very resource demanding, tightly coupled and often short lived.
```

> While advanced technices, like cluster autoscaling, can help with some of these challenges, there is a clear need for Kubernetes primitives representing higher level concepts that AI researches and data scientists are more familiar with.

Reviewer 2: ```suggestion
While advanced techniques, like cluster autoscaling, can help with some of these challenges, there is a clear need for Kubernetes primitives representing higher level concepts that AI researchers and data scientists are more familiar with.
```

### Why Kubeflow

This is where [Kubeflow](https://www.kubeflow.org/docs/started/introduction/) enters the picture.
Originally developed by Google, Kubeflow is, essentially, a collective of open-source projects that each provide a solution for different steps in a machine-learning workflow.
However, Kubeflow also positions itself as a complete platform, integrating all of those projects into a cohesive and easy to use solution while providing additional features and security.

Kubeflow makes it easy to train and serve models, supporting most of the commonly used machine learning frameworks.

> It also helps researches to reify their machine-learning workflows using "pipelines", leading to faster and more reproducible model development cycles.

Reviewer 2: ```suggestion
It also helps researchers to reify their machine-learning workflows using "pipelines", leading to faster and more reproducible model development cycles.
```

### Kubeflow Platform Architecture

> If all of that sounds to you like a very complex system with a lot of moving parts, for better of for worse, you are right.

Reviewer 2: ```suggestion
If all of that sounds to you like a very complex system with a lot of moving parts, for better or for worse, you are right.
```

> Kubeflow is composed of several components and is built on a base of quite a few third-party pojects:

Reviewer 2: ```suggestion
Kubeflow is composed of several components and is built on a base of quite a few third-party projects:
```

   - “Enable nodes on spot VMs” checked (recommended)

2. Add a second node pool using the button in the top bar (let's call this one `gpu`) and configure it with

   - “Number of nodes (per zone)” set to 0 (nodes in this group should only be provisioned on-demand)
   - “Enable cluster autoscaler” checked
   - “Size limit type” set to “Total limits”
   - “Minimum number of nodes” set to 0
   - “Maximum number of nodes” set to 3 (or more, depending on anticipated load)
   - “Machine configuration family” set to “GPUs”
   - “GPU type” set to L4 and “Number of GPUs” set to 1
   - “GPU Driver installation” set to “Google-managed” and driver version set to “Latest”
   - Machine type `g2-standard-4` (or use a different one if desired)
   - “Enable nodes on spot VMs” checked (again, recommended)
   - Under “Security” on the left navigation, select “Set access for each API” and change “Storage” to “Read Write”.

To provision your cluster, click “Create” on the bottom of the page and wait for your cluster to be ready.

:::note
If you get an error like “accelerator l4 is not available in region X”, go back to the “Cluster basics” page and check “Specify default node locations”.

> Then check all the regions except for the one mentionen in the error message and try creating the cluster again.

Reviewer 2: ```suggestion
Then check all the regions except for the one mentioned in the error message and try creating the cluster again.
```

Note down the IP address and use it to create a DNS record for this Kubeflow installation.
The exact procedure for this is obviously different for each domain registrar, but in the end, your record should look something like this:

```
kf.example-company.com. 60 IN A <your ingress IP address>
```

It will take some time for GCP to register all LoadBalancer as healthy and request the TLS certificate from the CA.

:::tip[Try it out]
Once everything is done, enter the domain in your browser and you will be able to log in, just like before.
:::

Since our Kubeflow deployment is now reachable for the public, you will probably want to set up a more robust authentication flow than the built-in default username and password.
Of course, you can do this by extending the `staticPasswords` list located in the Dex configuration, but where Dex really shines at is delegating authentication to an external Identity Provider IdP), such as Google, GitHub, Microsoft or basically any provider that implements the OIDC protocol.
So let's do that!

### Configuring External Authentication

> Like many other organizations, we use several of Microsofts producs, so it makes sense for us to use Azure Active Directory (AAD) here.

Reviewer: Microsofts **products**

To do this, add the following file:

```yaml title="custom/centraldashboard-deployment-patch.yaml"
apiVersion: apps/v1
kind: Deployment
metadata:
  name: centraldashboard
  namespace: kubeflow
spec:
  template:
    spec:
      containers:
        - name: centraldashboard
          env:
            - name: REGISTRATION_FLOW
              value: 'true'
```

Don't forget to add this patch to your `kustomization.yaml`:

```yaml title="custom/kustomization.yaml (final version)"
resources:
  - ../example
  - ingress
components:
  - microsoft-auth
patches:
  - path: istio-ingressgateway-service-patch.yaml
  # here we enable the registration flow feature
  - path: centraldashboard-deployment-patch.yaml
configMapGenerator:
  - name: oauth2-proxy
    namespace: oauth2-proxy
    files:
      - oauth2_proxy.cfg
    behavior: merge
```

And apply the changes by running `kustomize build custom | kubectl apply -f -`.

Once the command has completed, reload the page in your browser and you should be presented with the Kubeflow registration flow.
Complete it and you will see that you are now “owner” of a namespace!

## Final Thoughts

> This concludes our guide on how to setup vanilla Kubeflow on GCP.

Reviewer 2: ```suggestion
This concludes our guide on how to set up vanilla Kubeflow on GCP.
```
